### PR TITLE
fix(automation): refuse Hz passed to the MHz tune verb

### LINKED ISSUE #4550: Automation bridge: tune returns ok:true for a frequency it did not apply (Hz vs MHz)
## What

The automation bridge's `tune` verb returns `"ok": true` for a frequency it did not apply. Given the value in **Hz**, an out-of-band tune is silently discarded; the same target in **MHz** works.

## Reproduction

Identical start, identical target, only the unit differs:

```
park at 21.2 MHz
tune 14200000  ->  {"ok": true, "tune": 14200000}   slice stays 21.2   <-- no-op
park at 21.2 MHz
tune 14.2      ->  {"ok": true, "tune": 14.2}       slice becomes 14.2 <-- applied
```

More of the same:

```
tune 7100000   ->  ok=true, reply 7100000   slice unchanged at 14.3
tune 7.1       ->  ok=true, reply 7.1       slice 7.1
tune 21.2      ->  ok=true, reply 21.2      slice 21.2
```

Within the current band the Hz form appears to work — `tune 14250000` while already on 20 m does land on 14.25. That is what makes this easy to miss: it looks unit-agnostic until a band change, then it stops moving.

## What seems to be happening

The argument is consumed as MHz. `14200000` is read as 14,200,000 MHz, fails a range check somewhere, and the failure is not propagated — the verb still answers `ok: true`. The in-band case survives because the slice is clamped back into the current passband rather than rejected.

The echoed `tune` field reports the **raw input**, not the value applied, so the reply looks like confirmation. In the failing case above it says `14200000` while the radio sits on 21.2 MHz.

`docs/automation-bridge.md` does not state the unit for this verb.

## Why it matters

An automation client cannot tell success from failure. Every observable signal — `ok: true`, an echoed frequency, no error — says the tune happened. Only a follow-up `get slices` reveals it did not, and a script that trusts the reply will happily go on to key on the wrong band.

I hit this while scripting a calibration sweep: the harness tuned, confirmed `ok`, and transmitted, with the radio still on the previous band. Reading back the frequency after every tune is the workaround, but the reply should not claim success it cannot support.

## Suggested shape

Either is fine and they are not exclusive:

- Reject out-of-range input with `ok: false` and a reason, rather than returning success.
- Echo the frequency that was **actually applied** (post-clamp, in a stated unit), so a client can compare intent against result.

Documenting the expected unit in `docs/automation-bridge.md` would also help; accepting both and normalising would be friendlier still, given the in-band case already behaves as though both are accepted.

## Possibly related

The QSO recorder names files from a frequency that can be stale or wrongly scaled — during this session it produced `2026-07-28_153625Z_10.000MHz_USB.wav` while the radio was on 14.200 MHz for the whole session. I have not traced whether that shares a root cause; flagging it only because it looks like the same value being read in the wrong unit.

## Environment

**v26.7.4.1** (released build, version read from the running app via `whoami`), Windows 11, real Hermes-Lite 2 (gateware 74) and FLEX-6700, single slice.

### PR DESCRIPTION
**Pull request by nigelfenton**
Fixes #4550.

> **Revised after review.** This PR originally also tried to report what the
> slice actually took. That half was unreachable in the shipping app *and*
> could not have detected a rejection where it did run — it has been removed
> rather than moved. Detail in
> [this comment](https://github.com/aethersdr/AetherSDR/pull/4552#issuecomment-5111909430);
> the reasoning is preserved below.

## The bug

`tune` answers `ok: true` for a frequency it did not apply.

```
park at 21.2 MHz
tune 14200000  ->  {"ok": true, "tune": 14200000}   slice stays 21.2   <-- no-op
tune 14.2      ->  {"ok": true, "tune": 14.2}       slice becomes 14.2
```

The way in is passing Hz to a verb documented as `tune <mhz>`. The value is far outside any band, the radio ignores it, and every observable signal says the tune happened: `ok: true`, the frequency echoed, no error. Only a follow-up `get slices` reveals it. An in-band Hz value gets clamped and appears to work, so the mistake can stay hidden until the first band change — at which point a script tunes, confirms `ok`, and transmits on the previous band. That is how I found it, while automating a calibration sweep.

## The change

Refuse the unit mistake, naming the likely intent:

```
tune takes MHz, not Hz — got 14200000 (did you mean 14.200000?)
```

Deliberately **not** auto-converted — guessing would make `14200000` mean 14.2 MHz here and something else in every other frequency verb.

The ceiling is `105000` MHz: ten times the top of the band table (`BandDefs.h` — 3cm ends at 10500). That clears any real transverter setup by an order of magnitude while still catching the whole family of Hz-for-MHz mistakes, including `tune 500000` for 500 kHz, which a looser threshold lets through into the same silent no-op.

The guard sits **above** the `m_tuneHandler` dispatch, so it applies on the path the shipping app actually takes.

### What this PR does *not* fix

`ok: true` remains an acknowledgement, not a confirmation. `SliceModel::setFrequency()` records the tune optimistically and the radio's own value arrives asynchronously in `applyStatus()`, so there is nothing to read back synchronously. A request the radio rejects or clamps for some *other* reason still returns `ok: true` with the request echoed.

Closing that properly means waiting on the status update, in `MainWindow::automationTune` (`MainWindow.cpp:6355`) rather than here. It is a real change and deserves its own PR. Both the code and the bridge docs now say so plainly instead of implying the reply is authoritative.

## Verification — on air, real hardware

Hermes-Lite 2 `00:1C:C0:A2:13:DD`, gateware 74. The slice is parked at 21.2 MHz first, so a silent no-op shows up as the slice failing to move:

```
park at 21.2 MHz

tune 14200000  ->  ok:false  "tune takes MHz, not Hz — got 14200000 (did you mean 14.200000?)"
                   freq 21.2 -> 21.2   UNMOVED

tune 500000    ->  ok:false  "tune takes MHz, not Hz — got 500000 (did you mean 0.500000?)"
                   freq 21.2 -> 21.2   UNMOVED

tune 10368.1   ->  ok:true   freq 21.2     -> 10368.1   (3cm, top of the band table)
tune 14.2      ->  ok:true   freq 10368.1  -> 14.2
tune 7.1       ->  ok:true   freq 14.2     -> 7.1       (band change)
tune 14.25     ->  ok:true   freq 7.1      -> 14.25
```

Both refusals leave the slice where it was, which is the whole point — the old behaviour answered `ok:true` and moved nothing. `500000` is the case a looser threshold lets through: under 1e6 it fell past the guard and became a 500000 MHz request, landing back in the same silent no-op. `10368.1` confirms the ceiling refuses nothing real.

No regression for valid input, including across bands.

**Tests.** `automation_json_id_test` covers both refusals and pins that `10368.1` still tunes. That test installs a tune handler exactly as `MainWindow_Session.cpp:1937` does, so the cases also pin that the guard runs *before* the dispatch — they assert the handler was never reached. Verified red-before/green-after by restoring the looser threshold: the 500 kHz case fails, the rest pass.

`ctest -R "automation|bridge|slice"` — **16/16**, including `bridge_docs_check`. Full tree builds on Windows/MSVC.

## Note for reviewers

While investigating I assumed the stranded panadapter after a bridge tune was part of this. It is not — `tune` moves the slice only and `targettune` is the verb that recentres the pan, which is documented and working as intended. That observation is withdrawn.

`targettune` shares the same echo shape and the same optimistic-model caveat, and I have only exercised it with valid in-band values. It likely warrants both the guard and the async confirmation; I have not changed it here.

### CODE REVIEW

**Comment by nigelfenton** [comment]
Addressed in `5c5bba77`. You were right on both counts, and the conclusion is that the read-back had to go rather than move.

**It never ran.** Confirmed `MainWindow_Session.cpp:1937` installs the tune handler unconditionally during session setup, so `doTune()` returns at the `m_tuneHandler` dispatch (`AutomationServer.cpp:6127`) in the shipping app. The only callers reaching the block below are tests and a headless bridge with no MainWindow.

**And it could not have failed if it had.** `SliceModel::setFrequency()` assigns `m_frequency = mhz` before sending `slice tune`, so `s->frequency()` on the next line returns the request by construction. `qFuzzyCompare(applied, mhz)` was always true and the `requested`/`applied:false` branch was unreachable. The no-op I originally observed is the radio rejecting the tune and pushing a status update back milliseconds later, which is a different mechanism entirely from the one I wrote code against.

So the block is removed. Doing this properly means waiting on that asynchronous `applyStatus()` update, and doing it in `MainWindow::automationTune` (`MainWindow.cpp:6355`) where the app would actually see it — a real change, not a relocation, and not this PR. I've left a comment at the call site recording that the echo is an acknowledgement rather than a confirmation, and why, so the next person doesn't re-derive it and reach for the same wrong fix.

**The 500 kHz gap is closed.** You were right that `tune 500000` fell *under* the 1 THz threshold and became a 500000 MHz request — the same silent no-op, an octave lower. The ceiling is now `105000` MHz: ten times the top of the band table (`BandDefs.h`, 3cm ends at 10500), so it catches the whole family while clearing any real transverter setup by an order of magnitude.

**Docs updated** as you suggested — the refusal now sits in the `tune` section next to the existing refusal list, with the worked `14200000` example. I also documented the `ok:true` caveat there, since removing the read-back means that behaviour is now load-bearing for callers: the reply echoes the request, and `get slices` is how you find out where the radio landed.

### Test

`automation_json_id_test` covers both refusals and pins that `10368.1` — the 3cm calling frequency, top of the band table — still tunes, so the ceiling can't be tightened later without someone noticing.

Worth noting that test already installs a tune handler exactly as the app does, so the new cases also pin the ordering that made the guard survive this PR: they assert the handler was **never reached**. If the guard ever drifted below the dispatch it would go silently dead the same way the read-back did, and these would catch it.

Red-before/green-after: with the threshold back at `1e6`, the 500 kHz case fails and the others pass. Full tree builds; `ctest -R "automation|bridge|slice"` 16/16 on Windows/MSVC, including `bridge_docs_check`.

**Comment by nigelfenton** [comment]
The `targettune` follow-up from the reviewer note is now open as #4568, **stacked on this branch** — it hoists the guard into one shared helper (`refuseIfHzNotMhz`) and applies it to `targettune` and, as it turns out, `pan center` too.

`pan center` was the surprise: `RadioModel::setPanCenter()` clamps only the LOW edge, so an Hz-sized centre was optimistically stored **and advertised over TCI (`dds:`)** while the radio rejected it. Its own comment describes that behaviour; nothing upstream refused the value.

Merging this PR first makes #4568 a clean 3-file diff.

**Review by ten9876 — CHANGES_REQUESTED** [review]
The core change is sound and I checked the things most likely to be wrong before looking for problems: the guard sits **above** the `m_tuneHandler` dispatch so it covers the shipping GUI path, `/utf-8` is set for MSVC so the em-dash literal is safe, the doc prose sits inside the `tune` section without disturbing the GENERATED block (`gen_bridge_docs.py` checks that plus duplicate headings), the `[get slices](#get)` anchor resolves, and the new tests genuinely pin handler-not-reached ordering. The reasoning in the comment for *why* 105 GHz rather than 1 THz is good and I'd keep it.

One blocker.

## Blocker — `nan` passes both guards and reaches the radio

I compiled a Qt6 test against the exact conditions rather than reasoning about the parser:

```
nan      ok=1 val=nan  isfinite=0   mhz<=0: false   mhz>105000: false   → PASSES
NaN      ok=1 val=nan  isfinite=0   mhz<=0: false   mhz>105000: false   → PASSES
inf      ok=1 val=inf              mhz<=0: false   mhz>105000: true    → caught
-inf     ok=1 val=-inf             mhz<=0: true                        → caught
1e400    ok=0                                                          → caught
```

`QString::toDouble("nan")` returns **ok=true**, and every comparison against NaN is false — so neither `mhz <= 0` nor `mhz > kMaxTunableMhz` fires. `tune nan` reaches `s->setFrequency(mhz)` and sends `slice tune 0 nan` to the radio.

`inf` *is* caught, so this is the one remaining hole in the family rather than a general gap. Fix is one line folded into the existing check:

```cpp
    if (!okF || !std::isfinite(mhz) || mhz <= 0)
        return err(QStringLiteral("tune requires a positive frequency in MHz"));
```

Blocking because this is precisely the boundary this PR exists to validate, and it's the Constitution's boundary-input-validation principle applied to the verb the PR is named after.

## Worth doing in the same pass — and it affects #4568

I checked #4568, which refactors this rule into a shared `refuseIfHzNotMhz(verb, mhz)`. **The polarity flip changes the NaN behaviour by accident.** Here the test is `if (mhz > kMaxTunableMhz) return err(...)`, so NaN falls through and is accepted. There it is:

```cpp
if (mhz <= kMaxTunableMhz)
    return std::nullopt;
```

`NaN <= 105000` is also false, so NaN does *not* get the early return and ends up **refused** — but with a nonsense message: `takes MHz, not Hz — got nan (did you mean nan?)`.

So neither PR handles NaN deliberately: this one lets it through, that one rejects it while misdiagnosing it. Since #4568 is the version that survives, the `!std::isfinite` check belongs in that shared helper (or in each verb's positive-frequency check) so the message is right for the actual problem. Fixing #4552 first and letting #4568 inherit it is the cheaper order.

## Not blocking

**kHz-for-MHz still passes.** `tune 14200` is above the band table's 10.5 GHz top but under the 105 GHz ceiling, so it's accepted and the slice never moves — the same "script sees ok, then transmits on the old band" shape the PR is written against. I read the rest of `doTune` and the `ok:true`-on-no-op is *documented*, with a rationale, and explicitly scoped out:

> The reply echoes the REQUEST, and cannot do better from here … Closing the gap properly means waiting on that status update — a different change.

So I'm not treating that as a defect. The narrower question is about the guard's own stated job — catching unit mistakes that would otherwise silently no-op. kHz-for-MHz is that family, and the ceiling is already derived from the band table (10× its top), so consulting the table directly, or tightening the multiplier, would catch it for about the same code. Your call whether that's in scope for this PR or the next one; I'd just rather it be a decision than an omission.

**The 105 GHz ceiling refuses the real 122/134/241 GHz allocations** with no override, and XVTR RF frequency is user-configurable elsewhere in the app. Low practical impact — but the misdiagnosis is the worse half. Telling someone tuning 122 GHz "did you mean 0.122250?" is confidently wrong, which is the failure mode the rest of this guard's messaging avoids so carefully. Cheaper to special-case the message than to move the ceiling.

**Nit:** the comment cites `MainWindow.cpp:6355` for `automationTune`; it's 6322 on this head, and main has moved again since. Worth naming the function without the line number so it can't rot.

Thanks — the guard is the right idea and the messaging is unusually good. It's the NaN edge that needs closing before it goes in.

— Claude (AI dev partner)

**Comment by nigelfenton** [comment]
All addressed in `e8a58bdd`. The NaN table you compiled was the whole case — thanks for testing the parser instead of reasoning about it; I'd have got that wrong.

**Blocker — non-finite input.** The positive-frequency check now requires `std::isfinite`, so `nan`/`NaN`/`inf`/`-inf` are refused there with that check's own message ("tune requires a positive finite frequency in MHz") rather than falling through to a unit diagnosis that doesn't apply. Fixed in this PR so #4568's helper inherits it on rebase, per your suggested order.

**The 122/134/241 GHz misdiagnosis.** Taken. Ceiling-to-300-GHz now presents both readings — "if that was Hz, resend as 0.122250; if it is a millimetre-wave frequency in MHz, it is beyond what AE can tune" — and never says "did you mean". True Hz mistakes (>300 GHz as MHz, which covers every value the existing tests pin) keep the established message and prefix.

**kHz-for-MHz.** Decided, and now decided *visibly*: it passes, because `14200` is indistinguishable from a legitimate transverter request in exactly the headroom the ceiling exists to protect, and a band-table lookup would refuse real 24/47/76 GHz operation (XVTR RF is user-configurable beyond the table). The ceiling comment documents the reasoning and a test pins the acceptance so it's a tripwire, not an omission.

**Nit.** The rotting `MainWindow.cpp:6355` is now a function name.

Verification, both ways: the 4 non-finite cases and the 122.25 GHz message case all FAIL against the previous guard (confirmed by stash-rebuild-run), 22/22 PASS with the fix, handler-not-reached pinned throughout.

#4568 rebase with the helper inheriting all of this is next.

File: src/core/AutomationServer.cpp [hunk 1]
```diff
@@ -6102,8 +6103,53 @@ QJsonObject AutomationServer::doTune(const QString& value, const QString& id)
...
+        // transverter "did you mean 0.122250?" would be confidently wrong,
+        // which is the failure mode this guard's messaging exists to avoid.
+        if (mhz <= 300'000.0)
+            return err(QStringLiteral("tune takes MHz — got ")
+                       + QString::number(mhz, 'f', 0)
+                       + QStringLiteral(", above the 105000 MHz ceiling. If that was "
+                                        "Hz, resend as ")
+                       + QString::number(mhz / 1.0e6, 'f', 6)
+                       + QStringLiteral("; if it is a millimetre-wave frequency in "
+                                        "MHz, it is beyond what AE can tune"));
```
Comment (ten9876) [inline src/core/AutomationServer.cpp]: **Nit (would like fixed here) — the message contradicts itself just above the ceiling, and "AE" is not the app's name.**

`QString::number(mhz, 'f', 0)` rounds the value to whole MHz while the ceiling is spelled out at the same width, so anything in (105000, 105000.5) reports as *"got 105000, above the 105000 MHz ceiling"*. Probed on a local build at this head:

```
PROBE 105000.4  ok=0  err=tune takes MHz - got 105000, above the 105000 MHz ceiling. If that was Hz, resend as 0.105000; ...
```

Narrow, but this is the branch you added specifically so the guard would stop asserting things that aren't true, and telling someone their 105000.4 is above 105000 is that failure mode in miniature.

Two smaller things folded into the same edit: `105000` is written out here as well as in `kMaxTunableMhz` two lines up, so the two can drift; and `AE` never names the application in user-facing text anywhere in the tree — in the UI it means the NR2 **AE filter** (`MainWindow.cpp:8353`, `AetherDspWidget.cpp:669`, `SliceTroubleshootingDialog.cpp:817`). An automation client reading "beyond what AE can tune" has no way to connect that to AetherSDR.

```suggestion
            return err(QStringLiteral("tune takes MHz — got ")
                       + QString::number(mhz, 'g', 12)
                       + QStringLiteral(", above the ")
                       + QString::number(kMaxTunableMhz, 'f', 0)
                       + QStringLiteral(" MHz ceiling. If that was Hz, resend as ")
                       + QString::number(mhz / 1.0e6, 'f', 6)
                       + QStringLiteral("; if it is a millimetre-wave frequency in "
                                        "MHz, it is beyond what AetherSDR can tune"));
```

Keeps both existing assertions green (`millimetre-wave` present, `did you mean` absent). The `'f', 0` rounding is on the `> 300000` branch below too, though there it can only shave a fraction off a value that is wrong by six orders of magnitude, so it doesn't mislead the same way.

File: src/core/AutomationServer.cpp [hunk 2]
```diff
@@ -6102,8 +6103,53 @@ QJsonObject AutomationServer::doTune(const QString& value, const QString& id)
...
+    // the ones a 1 THz threshold let through, such as `tune 500000` for 500 kHz,
+    // which would otherwise fall past the guard and land back in the silent
+    // no-op this refusal exists to prevent.
+    //
+    // kHz-for-MHz (e.g. `tune 14200` for 20m) deliberately PASSES: that value
+    // is indistinguishable from a legitimate microwave request (14.2 GHz is
+    // inside the transverter headroom this ceiling exists to protect), and
+    // refusing the range would break real 24/47/76 GHz operation. A band-table
+    // lookup was considered and rejected for the same reason — XVTR RF
+    // frequency is user-configurable beyond the table. Decided, not overlooked.
```
Comment (ten9876) [inline src/core/AutomationServer.cpp:6164]: **Non-blocking — the 76 GHz half of this justification is already refused elsewhere in the app.**

I went looking for whether the headroom argument holds and found the app has a second, tighter answer to the same question:

```cpp
// src/gui/VfoWidget.cpp:1034  (and RxApplet.cpp:613, identical)
const double maxMhz = (onXvtr || highExplicitMhzEntry) ? 50000.0 : 54.0;
```

Typed frequency entry caps at **50 GHz** on XVTR. So 76 GHz cannot be dialled in the GUI today, and 122/134/241 GHz — the allocations the branch below is careful not to misdiagnose — are outside it by a wide margin. 24 and 47 GHz are fine either way.

Confirmed by probe that the bridge is now the looser of the two:

```
PROBE 50001   ok=1 handlerHit=1
PROBE 76000   ok=1 handlerHit=1
```

Not asking you to move the ceiling — 105000 is defensible and the reasoning for choosing it over 1 THz is good. But there are now two "maximum tunable frequency" constants a factor of two apart with nothing pointing at each other, and this comment argues for the looser one using a band the tighter one already rejects. A sentence naming `VfoWidget.cpp`'s `50000.0` and saying why the bridge is deliberately more permissive would keep the next person from treating either number as the authority.

File: src/core/AutomationServer.cpp [hunk 3]
```diff
@@ -6102,8 +6103,53 @@ QJsonObject AutomationServer::doTune(const QString& value, const QString& id)
         return err(QStringLiteral("no radio model available"));
     bool okF = false;
     const double mhz = value.toDouble(&okF);
-    if (!okF || mhz <= 0)
-        return err(QStringLiteral("tune requires a positive frequency in MHz"));
+    // toDouble() accepts "nan" and "inf". NaN in particular sails past every
+    // range check below (NaN <= 0 and NaN > ceiling are both false), so
+    // non-finite input must be refused here, by name, where the message is
+    // about the value itself rather than a unit mistake it isn't.
+    if (!okF || !std::isfinite(mhz) || mhz <= 0)
```
Comment (ten9876) [inline src/core/AutomationServer.cpp:6124]: **Non-blocking — the guard is one-sided; there is no floor.**

```
PROBE 1e-300   ok=1  handlerHit=1
```

`tune 1e-300` passes both checks and reaches the handler, where it becomes the same silent no-op with `ok:true` that this PR exists to eliminate — just at the other end. The GUI applies a floor on the same value (`VfoWidget.cpp:1058`: `freqMhz >= 0.001 && freqMhz <= maxMhz`), so there is an existing project answer to borrow.

The symmetric unit mistake is real, if rarer than the Hz one: GHz-for-MHz, `tune 0.0142` meaning 14.2 MHz, lands at 14.2 kHz and no-ops. Outside what #4550 reported, so I'd take it in #4568 with the shared helper rather than here — flagging it only so the helper gets designed with both ends in mind instead of inheriting the asymmetry.

File: docs/automation-bridge.md [hunk 4]
```diff
@@ -1077,6 +1077,25 @@ belongs to another client` when another client owns it (Multi-Flex). To
...
+
+```json
+→ {"cmd":"tune","value":"14200000"}
+← {"ok":false,"error":"tune takes MHz, not Hz — got 14200000 (did you mean 14.200000?)"}
+```
+
+The threshold is 105000 (ten times the top of the band table), so any value a
+transverter could plausibly need still passes. It is deliberately not
+auto-corrected: `14200000` would have to mean 14.2 MHz here and something else
+in every other frequency verb.
```
Comment (ten9876) [inline docs/automation-bridge.md:1090]: **Nit — one of the three refusal messages is documented.**

`doTune` can now answer with three distinct errors, and this section describes one. The bridge doc is the contract automation clients are written against, and a client matching on error text (or a human reading it after a script stops) will not find the other two here:

```suggestion
The threshold is 105000 (ten times the top of the band table), so any value a
transverter could plausibly need still passes. It is deliberately not
auto-corrected: `14200000` would have to mean 14.2 MHz here and something else
in every other frequency verb.

Two neighbouring refusals share the check:

- Between the ceiling and `300000` the value is genuinely ambiguous — an Hz
  mistake, or a real millimetre-wave frequency correctly given in MHz — so the
  error offers both readings instead of asserting a conversion:
  `tune takes MHz — got 122250, above the 105000 MHz ceiling. If that was Hz,
  resend as 0.122250; if it is a millimetre-wave frequency in MHz, it is beyond
  what AetherSDR can tune`.
- `nan`, `inf` and `-inf` parse as numbers but are refused with `tune requires a
  positive finite frequency in MHz`.
```

(Wording assumes the `AE` → `AetherSDR` suggestion on `AutomationServer.cpp` is taken; adjust if not.)

File: tests/automation_json_id_test.cpp [hunk 5]
```diff
@@ -154,6 +162,75 @@ int main(int argc, char** argv)
     expectRejected(tuneRequest(0.9999999), integerError,
                    "near-integer numeric id is not rounded down to a slice");
 
+    // ---- #4550: Hz passed to an MHz verb is refused, not silently no-opped ----
+    //
+    // A tune handler IS installed above, exactly as MainWindow_Session.cpp:1937
+    // does in the shipping app. That matters: the guard is only worth anything
+    // if it runs BEFORE the m_tuneHandler dispatch, so these cases also pin that
+    // ordering — expectRejected asserts the handler was never reached.
```
Comment (ten9876) [inline tests/automation_json_id_test.cpp]: **Nit — same line-number rot you fixed in the .cpp comment, one file over.**

`MainWindow_Session.cpp:1937` is correct on this head but it is **2153** on current `main`, so this citation is stale before the PR merges. The `AutomationServer.cpp` comment now names the function instead of the line, which is the right treatment; this one kept the number.

```suggestion
    // A tune handler IS installed above, exactly as MainWindow_Session.cpp's
    // setTuneHandler(...) call does in the shipping app. That matters: the guard
    // is only worth anything if it runs BEFORE the m_tuneHandler dispatch, so
    // these cases also pin that ordering — expectRejected asserts the handler
    // was never reached.
```

**Review by ten9876 — COMMENTED** [review]
Rebuilt and re-ran this at head `e8a58bdd` on Linux/GCC rather than reading it: `automation_json_id_test` is **22/22**, and the mutation check reproduces independently — deleting `!std::isfinite(mhz)` turns `nan`, `NaN` and `inf` red (`-inf` survives on `mhz <= 0`), so the new cases pin the guard rather than the code. I also probed the verb directly with 22 boundary values; the table is at the end.

The NaN hole is closed, the mm-wave misdiagnosis is gone, and the kHz decision is now visible in a comment and a test instead of being an omission. All three of my previous points are addressed.

## Issue fit

#4550's repro closes exactly, on the path the shipping app takes. The guard sits above the `m_tuneHandler` dispatch and my probe confirms every refusal leaves `handlerCalls` unchanged, so the ordering the tests assert is real and not an artefact of the fixture. The issue's second suggested shape — echo the frequency actually applied — is deliberately not done, and the reason is now recorded in the code and in `docs/automation-bridge.md` rather than implied. Given the read-back was unreachable, recording the constraint is worth more than a relocated version of it.

Nothing here reads as smuggled preference: a refusal guard, its docs, and its tests. No defaults moved, no UI touched, no scope creep.

## No blockers

Six notes, none blocking. The first is the only one I'd like fixed in this PR, because it lands inside the branch that exists to prevent confidently-wrong messages.

1. **`tune 105000.4` answers "got 105000, above the 105000 MHz ceiling"** — the value is rounded to zero decimals while the ceiling is spelled out at the same width, so the message contradicts itself for anything in (105000, 105000.5). Suggestion inline (AutomationServer.cpp:6140).
2. **"AE" in a user-facing string.** Grepping the tree, `AE` never names the application in user-facing text — it means the NR2 "AE filter" (`MainWindow.cpp:8353`, `AetherDspWidget.cpp:669`). Folded into the same suggestion.
3. **The ceiling is hardcoded in the message text** as well as in `kMaxTunableMhz` right above it. Same suggestion derives it.
4. **The kHz rationale cites 76 GHz, which the app already refuses.** `VfoWidget.cpp:1034` and `RxApplet.cpp:613` cap typed entry at `50000.0` MHz on XVTR, so 76 GHz cannot be entered in the GUI today. My probe confirms `tune 50001` and `tune 76000` are accepted by the bridge — the bridge is now the more permissive of the two, by 2x, with no cross-reference between the constants (AutomationServer.cpp:6128).
5. **No lower bound.** `tune 1e-300` is accepted and reaches the handler (probed). The GUI requires `freqMhz >= 0.001` on the same value. The symmetric unit mistake — GHz-for-MHz, `tune 0.0142` for 14.2 MHz — still lands in the silent no-op this guard exists to prevent. Out of the reported scope; flagging it because the guard is one-sided (AutomationServer.cpp:6110).
6. **Docs cover one of the three refusals.** The Hz message is documented; the ambiguous-window message and the non-finite message are not. Suggestion inline (docs/automation-bridge.md:1087). Plus a stale `MainWindow_Session.cpp:1937` in the test comment — it is 2153 on current main, the same rot you fixed in the .cpp comment (tests/automation_json_id_test.cpp:167).

## One thing for the maintainer, not for you

`14200000` typed into the VFO field becomes 14.2 MHz — `VfoWidget.cpp:1029-1055` runs exactly the Hz/kHz heuristic this PR declines to run. So after this lands the GUI silently converts the value the bridge refuses. Your reasoning for refusing is sound and it is what the issue asked for, and I would not change it on my own judgement; but the two paths now disagree about what `14200000` means, and that is a coherence call above a reviewer's pay grade.

Separately: `Fixes #4550` will auto-close a `safety` / `priority: high` issue whose broader half — `ok:true` is an acknowledgement, not a confirmation — stays open with no tracking issue. #4568 covers `targettune` and `pan center`, not the async confirmation in `MainWindow::automationTune`. Worth opening that one before this merges so it does not fall off with the issue.

## Probe output (head `e8a58bdd`, local build)

```
value           ok  handler  error
14200000         0     0     tune takes MHz, not Hz - got 14200000 (did you mean 14.200000?)
500000           0     0     tune takes MHz, not Hz - got 500000 (did you mean 0.500000?)
122250           0     0     tune takes MHz - got 122250, above the 105000 MHz ceiling...
105000           1     1
105000.4         0     0     tune takes MHz - got 105000, above the 105000 MHz ceiling...   <-- note 1
105001           0     0     tune takes MHz - got 105001, above the 105000 MHz ceiling...
300000           0     0     tune takes MHz - got 300000, above the 105000 MHz ceiling...
300001           0     0     tune takes MHz, not Hz - got 300001 (did you mean 0.300001?)
50001            1     1                                                                    <-- note 4
76000            1     1                                                                    <-- note 4
nan              0     0     tune requires a positive finite frequency in MHz
inf              0     0     tune requires a positive finite frequency in MHz
-inf             0     0     tune requires a positive finite frequency in MHz
1e400            0     0     tune requires a positive finite frequency in MHz
0                0     0     tune requires a positive finite frequency in MHz
-5               0     0     tune requires a positive finite frequency in MHz
1e-300           1     1                                                                    <-- note 5
"  14.2  "       1     1
14.2abc          0     0     tune requires a positive finite frequency in MHz
1.42e7           0     0     tune takes MHz, not Hz - got 14200000 (did you mean 14.200000?)
0x1p10           0     0     tune requires a positive finite frequency in MHz
14,200,000       0     0     tune requires a positive finite frequency in MHz
```

Good fix, and the comments carry their reasoning well enough that the next person will not have to re-derive any of it.

--- Claude (AI dev partner)

**Review by ten9876 — APPROVED** [review]
## Approving — all review points resolved in `57bb4b6b`

I pushed the fixes to this branch rather than sending you round again; every item from my last pass is closed. Built and tested locally on Linux/GCC at the new head, not just read.

**Fixed**

1. **Rejected values are quoted back unrounded.** `'f', 0` rounded the value to whole MHz, so anything in (105000, 105000.5) reported as *"got 105000, above the 105000 MHz ceiling"*. A `formatMhz()` helper next to `err()` now renders with `'g'`/15 — plain notation for every realistic value, no rounding, and no 300-digit expansion for a tiny one. Placed next to `err()` deliberately so #4568's `refuseIfHzNotMhz()` inherits it on rebase instead of growing a second copy.
2. **The ceiling is no longer written twice.** The message derives it from `kMaxTunableMhz` instead of repeating `105000` two lines below the constant.
3. **`AE` → `AetherSDR`** in the user-facing string.
4. **Floor added** — `kMinTunableMhz = 0.001`, matching what typed frequency entry already applies to the same value. `tune 1e-300` previously passed both guards and reached `setFrequency()` and the radio. The comment is explicit that this is *not* a unit guard: GHz-for-MHz on HF lands at 14.2 kHz, a plausible VLF frequency rather than a diagnosable mistake, so the floor makes no claim to catch it.
5. **The kHz rationale no longer cites 76 GHz**, which typed entry already refuses (`VfoWidget.cpp` caps XVTR at `50000.0`). It now cites 24/47 GHz — both inside that cap — and records why the bridge ceiling is deliberately the looser of the two, so neither number reads as "the" tuning limit.
6. **Docs carry the full refusal table.** The ambiguous-window and non-finite messages were undocumented; a client matching on error text needed all four. The kHz-passes decision is documented there too, not only in the code.
7. **Both remaining line-number citations are function names now** (`SliceModel.cpp:117-124`, `MainWindow_Session.cpp:1937`). Both were accurate on this head and one `main` merge from rotting — 1937 is already 2153 upstream.

**Verification.** `automation_json_id_test` 25/25; `ctest -R "automation|bridge|slice"` 16/16 including `bridge_docs_check` and `gen_bridge_docs.py --check`. Red-before/green-after done the way you did it: restoring the rounding and deleting the floor fails exactly the two new cases and moves nothing else — including "kHz-for-MHz passes" and "top of the band table still tunes", which stay green as they should.

**Two things I did not change, both yours to call**

- **`Fixes #4550` will auto-close a `safety` / `priority: high` issue whose second half stays open.** #4568 covers `targettune` and `pan center`; nothing tracks the async confirmation in `MainWindow::automationTune`, and I could not find an existing issue for it. Your write-up in this PR is the best statement of that problem anywhere — worth lifting into an issue before this merges so it does not close with #4550.
- **The GUI and the bridge now disagree about `14200000`.** `VfoWidget.cpp` runs exactly the Hz/kHz heuristic this PR declines to run, so the same digits typed into the VFO field become 14.2 MHz while the bridge refuses them. Refusing is the safer half and it is what the issue asked for; I left it alone because reconciling the two is a product call, not a review one.

The guard is right, the messaging is unusually careful, and the comments carry enough reasoning that the next person will not re-derive any of it. Nice work — and thanks for taking the NaN table seriously rather than arguing with it.

--- Claude (AI dev partner)
